In [83]:
import pandas as pd

In [170]:
df = pd.read_csv(r'E:\Internships\Internship_InfosysSpringboard\ImpactSense - Earthquake Impact Prediction\Milestone_4\Day_32\data\database.csv')
df.head()

Date      Time  Latitude  Longitude        Type  Depth  Depth Error  \
0  01/02/1965  13:44:18    19.246    145.616  Earthquake  131.6          NaN   
1  01/04/1965  11:29:49     1.863    127.352  Earthquake   80.0          NaN   
2  01/05/1965  18:05:58   -20.579   -173.972  Earthquake   20.0          NaN   
3  01/08/1965  18:49:43   -59.076    -23.557  Earthquake   15.0          NaN   
4  01/09/1965  13:32:50    11.938    126.427  Earthquake   15.0          NaN   

   Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0                     NaN        6.0             MW  ...   
1                     NaN        5.8             MW  ...   
2                     NaN        6.2             MW  ...   
3                     NaN        5.8             MW  ...   
4                     NaN        5.8             MW  ...   

   Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                         NaN            NaN                  NaN   
1                         NaN            NaN                  NaN   
2                         NaN            NaN                  NaN   
3                         NaN            NaN                  NaN   
4                         NaN            NaN                  NaN   

   Horizontal Error  Root Mean Square            ID  Source Location Source  \
0               NaN               NaN  ISCGEM860706  ISCGEM          ISCGEM   
1               NaN               NaN  ISCGEM860737  ISCGEM          ISCGEM   
2               NaN               NaN  ISCGEM860762  ISCGEM          ISCGEM   
3               NaN               NaN  ISCGEM860856  ISCGEM          ISCGEM   
4               NaN               NaN  ISCGEM860890  ISCGEM          ISCGEM   

  Magnitude Source     Status  
0           ISCGEM  Automatic  
1           ISCGEM  Automatic  
2           ISCGEM  Automatic  
3           ISCGEM  Automatic  
4           ISCGEM  Automatic  

[5 rows x 21 columns]

In [172]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23412 entries, 0 to 23411
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        23412 non-null  object 
 1   Time                        23412 non-null  object 
 2   Latitude                    23412 non-null  float64
 3   Longitude                   23412 non-null  float64
 4   Type                        23412 non-null  object 
 5   Depth                       23412 non-null  float64
 6   Depth Error                 4461 non-null   float64
 7   Depth Seismic Stations      7097 non-null   float64
 8   Magnitude                   23412 non-null  float64
 9   Magnitude Type              23409 non-null  object 
 10  Magnitude Error             327 non-null    float64
 11  Magnitude Seismic Stations  2564 non-null   float64
 12  Azimuthal Gap               7299 non-null   float64
 13  Horizontal Distance         160

In [173]:
(df.isnull().sum()/df.shape[0])*100

Date                           0.000000
Time                           0.000000
Latitude                       0.000000
Longitude                      0.000000
Type                           0.000000
Depth                          0.000000
Depth Error                   80.945669
Depth Seismic Stations        69.686486
Magnitude                      0.000000
Magnitude Type                 0.012814
Magnitude Error               98.603280
Magnitude Seismic Stations    89.048351
Azimuthal Gap                 68.823680
Horizontal Distance           93.148813
Horizontal Error              95.062361
Root Mean Square              25.884162
ID                             0.000000
Source                         0.000000
Location Source                0.000000
Magnitude Source               0.000000
Status                         0.000000
dtype: float64

In [174]:
df.columns

Index(['Date', 'Time', 'Latitude', 'Longitude', 'Type', 'Depth', 'Depth Error',
       'Depth Seismic Stations', 'Magnitude', 'Magnitude Type',
       'Magnitude Error', 'Magnitude Seismic Stations', 'Azimuthal Gap',
       'Horizontal Distance', 'Horizontal Error', 'Root Mean Square', 'ID',
       'Source', 'Location Source', 'Magnitude Source', 'Status'],
      dtype='object')

In [175]:
earth_df = df[['Latitude', 'Longitude', 'Type', 'Depth','Magnitude', 'Magnitude Type',
                'Root Mean Square','Status']]
earth_df.head()

Latitude  Longitude        Type  Depth  Magnitude Magnitude Type  \
0    19.246    145.616  Earthquake  131.6        6.0             MW   
1     1.863    127.352  Earthquake   80.0        5.8             MW   
2   -20.579   -173.972  Earthquake   20.0        6.2             MW   
3   -59.076    -23.557  Earthquake   15.0        5.8             MW   
4    11.938    126.427  Earthquake   15.0        5.8             MW   

   Root Mean Square     Status  
0               NaN  Automatic  
1               NaN  Automatic  
2               NaN  Automatic  
3               NaN  Automatic  
4               NaN  Automatic

In [176]:
earth_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23412 entries, 0 to 23411
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Latitude          23412 non-null  float64
 1   Longitude         23412 non-null  float64
 2   Type              23412 non-null  object 
 3   Depth             23412 non-null  float64
 4   Magnitude         23412 non-null  float64
 5   Magnitude Type    23409 non-null  object 
 6   Root Mean Square  17352 non-null  float64
 7   Status            23412 non-null  object 
dtypes: float64(5), object(3)
memory usage: 1.4+ MB


In [177]:
(earth_df.isnull().sum()/earth_df.shape[0])*100

Latitude             0.000000
Longitude            0.000000
Type                 0.000000
Depth                0.000000
Magnitude            0.000000
Magnitude Type       0.012814
Root Mean Square    25.884162
Status               0.000000
dtype: float64

In [178]:
earth_df['Root Mean Square'].sample(10)

1078      NaN
21422    0.75
3645      NaN
12683    0.90
2126      NaN
1627      NaN
318       NaN
12752    0.90
22506    0.82
23312    0.76
Name: Root Mean Square, dtype: float64

In [179]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

# Assuming 'earthquake' is your DataFrame

# Select features to predict "Root Mean Square" (drop non-numeric or encode if needed)
features = ['Latitude', 'Longitude', 'Depth', 'Magnitude']  # Use relevant numeric features

# Split data into known and missing for target column
train_df = earth_df[earth_df['Root Mean Square'].notna()]
test_df = earth_df[earth_df['Root Mean Square'].isna()]

# X and y for training
X_train = train_df[features]
y_train = train_df['Root Mean Square']

# X for prediction
X_test = test_df[features]

# Initialize and train regression model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predict missing values
predicted_values = model.predict(X_test)

# Impute missing values
earth_df.loc[earth_df['Root Mean Square'].isna(), 'Root Mean Square'] = predicted_values

print(earth_df['Root Mean Square'])

0        1.0284
1        1.1056
2        1.0350
3        1.0655
4        1.0101
          ...  
23407    0.1898
23408    0.2187
23409    1.5200
23410    1.4300
23411    0.9100
Name: Root Mean Square, Length: 23412, dtype: float64


In [180]:
(earth_df.isnull().sum()/earth_df.shape[0])*100

Latitude            0.000000
Longitude           0.000000
Type                0.000000
Depth               0.000000
Magnitude           0.000000
Magnitude Type      0.012814
Root Mean Square    0.000000
Status              0.000000
dtype: float64

In [181]:
earth_df.dropna(inplace=True)

C:\Users\babin\AppData\Local\Temp\ipykernel_12388\791176556.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  earth_df.dropna(inplace=True)


In [182]:
(earth_df.isnull().sum()/earth_df.shape[0])*100

Latitude            0.0
Longitude           0.0
Type                0.0
Depth               0.0
Magnitude           0.0
Magnitude Type      0.0
Root Mean Square    0.0
Status              0.0
dtype: float64

In [183]:
# Feature engineering
def create_damage_potential(mag, depth):
    return 0.6 * mag + 0.2 * (700 - depth) / 700 * 10

earth_df['Damage_Potential'] = create_damage_potential(earth_df.Magnitude, earth_df.Depth)

earth_df.head()

C:\Users\babin\AppData\Local\Temp\ipykernel_12388\824091105.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  earth_df['Damage_Potential'] = create_damage_potential(earth_df.Magnitude, earth_df.Depth)


Latitude  Longitude        Type  Depth  Magnitude Magnitude Type  \
0    19.246    145.616  Earthquake  131.6        6.0             MW   
1     1.863    127.352  Earthquake   80.0        5.8             MW   
2   -20.579   -173.972  Earthquake   20.0        6.2             MW   
3   -59.076    -23.557  Earthquake   15.0        5.8             MW   
4    11.938    126.427  Earthquake   15.0        5.8             MW   

   Root Mean Square     Status  Damage_Potential  
0            1.0284  Automatic          5.224000  
1            1.1056  Automatic          5.251429  
2            1.0350  Automatic          5.662857  
3            1.0655  Automatic          5.437143  
4            1.0101  Automatic          5.437143

In [184]:
earth_df['Type'].value_counts()

Type
Earthquake           23229
Nuclear Explosion      175
Explosion                4
Rock Burst               1
Name: count, dtype: int64

In [185]:
earth_df = earth_df[earth_df['Type'] == 'Earthquake']


In [186]:
earth_df['Type'].value_counts()

Type
Earthquake    23229
Name: count, dtype: int64

In [187]:
earth_df.drop('Type', axis=1, inplace=True)

In [188]:
earth_df

Latitude  Longitude   Depth  Magnitude Magnitude Type  \
0       19.2460   145.6160  131.60        6.0             MW   
1        1.8630   127.3520   80.00        5.8             MW   
2      -20.5790  -173.9720   20.00        6.2             MW   
3      -59.0760   -23.5570   15.00        5.8             MW   
4       11.9380   126.4270   15.00        5.8             MW   
...         ...        ...     ...        ...            ...   
23407   38.3917  -118.8941   12.30        5.6             ML   
23408   38.3777  -118.8957    8.80        5.5             ML   
23409   36.9179   140.4262   10.00        5.9            MWW   
23410   -9.0283   118.6639   79.00        6.3            MWW   
23411   37.3973   141.4103   11.94        5.5             MB   

       Root Mean Square     Status  Damage_Potential  
0                1.0284  Automatic          5.224000  
1                1.1056  Automatic          5.251429  
2                1.0350  Automatic          5.662857  
3                1.0655  Automatic          5.437143  
4                1.0101  Automatic          5.437143  
...                 ...        ...               ...  
23407            0.1898   Reviewed          5.324857  
23408            0.2187   Reviewed          5.274857  
23409            1.5200   Reviewed          5.511429  
23410            1.4300   Reviewed          5.554286  
23411            0.9100   Reviewed          5.265886  

[23229 rows x 8 columns]

In [189]:
earth_df['Magnitude Type'].value_counts()

Magnitude Type
MW     7722
MWC    5668
MB     3594
MWB    2458
MWW    1983
MS     1702
ML       65
MWR      26
MD        6
MH        5
Name: count, dtype: int64

In [190]:
from sklearn.preprocessing import OneHotEncoder

# Select the columns to encode
columns_to_encode = ['Magnitude Type', 'Status']

# Initialize the encoder (ignore unknown categories, output dense array, integer dtype)
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False, dtype=int)

# Fit and transform the data for the selected columns
encoded_array = encoder.fit_transform(earth_df[columns_to_encode])

# Convert the encoded array to a DataFrame with proper column names
encoded_df = pd.DataFrame(encoded_array, columns=encoder.get_feature_names_out(columns_to_encode))

# Concatenate with original DataFrame (dropping the original columns to encode)
earth_df = pd.concat([earth_df.drop(columns=columns_to_encode), encoded_df], axis=1)

# Display first few rows
earth_df.head()


Latitude  Longitude  Depth  Magnitude  Root Mean Square  Damage_Potential  \
0    19.246    145.616  131.6        6.0            1.0284          5.224000   
1     1.863    127.352   80.0        5.8            1.1056          5.251429   
2   -20.579   -173.972   20.0        6.2            1.0350          5.662857   
3   -59.076    -23.557   15.0        5.8            1.0655          5.437143   
4    11.938    126.427   15.0        5.8            1.0101          5.437143   

   Magnitude Type_MB  Magnitude Type_MD  Magnitude Type_MH  Magnitude Type_ML  \
0                0.0                0.0                0.0                0.0   
1                0.0                0.0                0.0                0.0   
2                0.0                0.0                0.0                0.0   
3                0.0                0.0                0.0                0.0   
4                0.0                0.0                0.0                0.0   

   Magnitude Type_MS  Magnitude Type_MW  Magnitude Type_MWB  \
0                0.0                1.0                 0.0   
1                0.0                1.0                 0.0   
2                0.0                1.0                 0.0   
3                0.0                1.0                 0.0   
4                0.0                1.0                 0.0   

   Magnitude Type_MWC  Magnitude Type_MWR  Magnitude Type_MWW  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   Status_Automatic  Status_Reviewed  
0               1.0              0.0  
1               1.0              0.0  
2               1.0              0.0  
3               1.0              0.0  
4               1.0              0.0

In [191]:
(earth_df.isnull().sum()/earth_df.shape[0])*100

Latitude              0.78165
Longitude             0.78165
Depth                 0.78165
Magnitude             0.78165
Root Mean Square      0.78165
Damage_Potential      0.78165
Magnitude Type_MB     0.78165
Magnitude Type_MD     0.78165
Magnitude Type_MH     0.78165
Magnitude Type_ML     0.78165
Magnitude Type_MS     0.78165
Magnitude Type_MW     0.78165
Magnitude Type_MWB    0.78165
Magnitude Type_MWC    0.78165
Magnitude Type_MWR    0.78165
Magnitude Type_MWW    0.78165
Status_Automatic      0.78165
Status_Reviewed       0.78165
dtype: float64

In [192]:
earth_df.dropna(inplace=True)

In [193]:
(earth_df.isnull().sum()/earth_df.shape[0])*100

Latitude              0.0
Longitude             0.0
Depth                 0.0
Magnitude             0.0
Root Mean Square      0.0
Damage_Potential      0.0
Magnitude Type_MB     0.0
Magnitude Type_MD     0.0
Magnitude Type_MH     0.0
Magnitude Type_ML     0.0
Magnitude Type_MS     0.0
Magnitude Type_MW     0.0
Magnitude Type_MWB    0.0
Magnitude Type_MWC    0.0
Magnitude Type_MWR    0.0
Magnitude Type_MWW    0.0
Status_Automatic      0.0
Status_Reviewed       0.0
dtype: float64

In [194]:
# Save the processed DataFrame to a new CSV file
earth_df.to_csv(r'E:\Internships\Internship_InfosysSpringboard\ImpactSense - Earthquake Impact Prediction\Milestone_4\Day_31\data\preprocessed_earthquake_data.csv', index=False)


In [200]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Read CSV
df = pd.read_csv("E:/Internships/Internship_InfosysSpringboard/ImpactSense - Earthquake Impact Prediction/Milestone_4/Day_32/data/database.csv")

# Split into 80/20
train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)

# Save
train_df.to_csv("E:/Internships/Internship_InfosysSpringboard/ImpactSense - Earthquake Impact Prediction/Milestone_4/Day_32/data/train_data.csv", index=False)
test_df.to_csv("E:/Internships/Internship_InfosysSpringboard/ImpactSense - Earthquake Impact Prediction/Milestone_4/Day_32/data/test_data.csv", index=False)


In [201]:
df = pd.read_csv('E:/Internships/Internship_InfosysSpringboard/ImpactSense - Earthquake Impact Prediction/Milestone_4/Day_32/data/processed_trained_data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23232 entries, 0 to 23231
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Latitude                23232 non-null  float64
 1   Longitude               23232 non-null  float64
 2   Depth                   23232 non-null  float64
 3   Magnitude               23232 non-null  float64
 4   Root Mean Square        23232 non-null  float64
 5   Damage_Potential        23232 non-null  float64
 6   Magnitude Type_MD       23232 non-null  float64
 7   Magnitude Type_MH       23232 non-null  float64
 8   Magnitude Type_ML       23232 non-null  float64
 9   Magnitude Type_MS       23232 non-null  float64
 10  Magnitude Type_MW       23232 non-null  float64
 11  Magnitude Type_MWB      23232 non-null  float64
 12  Magnitude Type_MWC      23232 non-null  float64
 13  Magnitude Type_MWR      23232 non-null  float64
 14  Magnitude Type_MWW      23232 non-null

# Summary of Data Preprocessing

This notebook demonstrates the complete data preprocessing pipeline for the ImpactSense earthquake prediction project. Here's what we accomplished:

## Key Steps:
1. **Data Loading**: Loaded raw earthquake data from database.csv
2. **Feature Selection**: Selected relevant columns for analysis
3. **Missing Value Handling**: Imputed missing Root Mean Square values using RandomForest
4. **Data Filtering**: Filtered to include only earthquake records
5. **Feature Engineering**: Created Damage_Potential feature based on magnitude and depth
6. **Categorical Encoding**: One-hot encoded categorical variables (Magnitude Type, Status)
7. **Data Splitting**: Split data into train/test sets
8. **Final Processing**: Created processed_trained_data.csv for model training

## Results:
- Original data shape: 23,412 records with 21 features
- Processed data shape: 23,232 earthquake records with 17 features
- Ready for machine learning model training

In [202]:
# Final verification: Check the processed data
import pandas as pd
df_processed = pd.read_csv('data/processed_trained_data.csv')

print("Final processed data statistics:")
print(f"Shape: {df_processed.shape}")
print(f"Columns: {list(df_processed.columns)}")
print("\nData types:")
print(df_processed.dtypes)
print("\nFirst few rows:")
print(df_processed.head())

# Check for any remaining missing values
print(f"\nMissing values per column:")
print(df_processed.isnull().sum())

# Verify Damage_Potential feature
print(f"\nDamage_Potential statistics:")
print(df_processed['Damage_Potential'].describe())

Final processed data statistics:
Shape: (23232, 17)
Columns: ['Latitude', 'Longitude', 'Depth', 'Magnitude', 'Root Mean Square', 'Damage_Potential', 'Magnitude Type_MD', 'Magnitude Type_MH', 'Magnitude Type_ML', 'Magnitude Type_MS', 'Magnitude Type_MW', 'Magnitude Type_MWB', 'Magnitude Type_MWC', 'Magnitude Type_MWR', 'Magnitude Type_MWW', 'Magnitude Type_Unknown', 'Status_Reviewed']

Data types:
Latitude                  float64
Longitude                 float64
Depth                     float64
Magnitude                 float64
Root Mean Square          float64
Damage_Potential          float64
Magnitude Type_MD         float64
Magnitude Type_MH         float64
Magnitude Type_ML         float64
Magnitude Type_MS         float64
Magnitude Type_MW         float64
Magnitude Type_MWB        float64
Magnitude Type_MWC        float64
Magnitude Type_MWR        float64
Magnitude Type_MWW        float64
Magnitude Type_Unknown    float64
Status_Reviewed           float64
dtype: object

First f